APE26.rst

```diff
@@ -0,0 +1,325 @@
+store and handle data has resulted in a large amount of code
+duplication (not just in the code proper, but also in the tests,
+leading to missed combinations). Restructuring the ``Frame``
+classes to remove data storage will allow for much more
+maintainable, de-duplicated code. It will also make it easier to
+contribute: if there is a problem one would like to solve in a
+given method, if one looks in ``SkyCoord``, one will likely find
+that it does not exist, but instead is defined on
+``BaseCoordinateFrame`` and gets dynamically called via
+``SkyCoord.__getattr__``. Indeed, the construction of
+``BaseCoordinateFrame`` ends up complicating ``SkyCoord``, which
+has to manage the coordinate data **through** the stored
+``BaseCoordinateFrame``.
+
+Another issue with the current implementation of coordinate
+frames is that the optional inclusion of coordinate data makes
+the reference frames “multi-modal”. This creates different usage
+modes (with and without data), each exhibiting different
+behavior. For instance, some methods such as ``separation`` work
+fine with frames with data, while doing this on coordinate frames
+without data results in an error. While this multi-modal
+structure as motivated in `APE 5
+<https://github.com/astropy/astropy-APEs/blob/main/APE5.rst>`_
+can be seen as a benefit for interpretability of the logic, we
+also now see it from the perspective of an anti-pattern. It
+forces any code interacting with the frame classes to handle both
+cases (checking ``frame.has_data``), complicating the codebase.
+Moreover static analyzers cannot determine which case is being
+used; with separated frames and data static analyzers will be
+able to prevent this entire class of errors. We thus find it
+worthwhile to separate coordinate data from reference frames at
+the possible expense of developers having to learn this new
+framework.
+
+All the points discussed thus far – separation of concerns and
+code duplication – concern maintainers. However user experience
+is the more important consideration. In this arena too,
+separating frames from data storage has its advantages. Perhaps
+most importantly, documentation will be more obvious: the methods
+and attributes are defined on ``SkyCoord`` (and ``Coordinate``)
+proper, so sphinx will know how to typeset those, while type
+checkers can help users in finding and using them properly. It
+will also be easier: following the Zen of Python, there should be
+one clear way to do something. The present overlap leads to
+confusion wherein beginner users end up creating
+``BaseCoordinateFrame`` objects such as ``ICRS``, when the docs
+are clear that these are for more advanced users and that
+``SkyCoord`` is to be preferred. The system will also be less
+fragile. For example, if users manipulate the internal workings
+of ``SkyCoord`` (which is discouraged but possible), the
+coordinate data can become decoupled from the caching that
+``SkyCoord`` performs for speed. With these proposed changes
```

Comment:
I am not worried about users who mess with `SkyCoord` internals. The problems they will cause will be their own fault. But it might be worth considering downstream developers who subclass `SkyCoord`.

Reply:
That is a better example to highlight.

Reply:
Updated the text accordingly in [062755f](https://github.com/astropy/astropy-APEs/pull/112/commits/062755fab13a96d06e511507412c04eee929b32e).